From 2fadb0ade68280a99611b570a5a93ab39e1b419f Mon Sep 17 00:00:00 2001
From: Bharath Ramesh <rameshb@vmware.com>
Date: Wed, 8 Oct 2014 17:17:11 -0700
Subject: [PATCH 2/7] Build pywin32 using vs2012

-- Modified setup.py to be compatible with vs2012

Reviewed by: akreher, arolett
Review URL: https://reviewboard.eng.vmware.com/r/680627/

Conflicts:
	setup.py
---
 com/win32comext/axdebug/src/stdafx.h |  9 ++++----
 setup.py                             | 42 +++++++++++++++++++++++++++++-------
 2 files changed, 39 insertions(+), 12 deletions(-)

diff --git a/com/win32comext/axdebug/src/stdafx.h b/com/win32comext/axdebug/src/stdafx.h
index 1ebc570..0ca9433 100644
--- a/com/win32comext/axdebug/src/stdafx.h
+++ b/com/win32comext/axdebug/src/stdafx.h
@@ -24,8 +24,7 @@ typedef BSTR UserBSTR;
 #include <activdbg.h>
 #else
 #include "activdbg.h"
-#endif
-
+// rameshb: Newer windows SDK's dbgprop.h has re-introduced these
 #if defined(__REQUIRED_RPCNDR_H_VERSION__)
 // for some strange reason, these no longer exist in dbgprop.h !?!?
 enum __MIDL___MIDL_itf_dbgprop_0000_0001
@@ -60,6 +59,10 @@ enum __MIDL___MIDL_itf_dbgprop_0000_0002
     };
 typedef DWORD DBGPROP_INFO_FLAGS;
 
+#endif
+#endif
+
+// rameshb: Newer windows SDK's dbgprop.h is yet to re-introduced these
 enum {
    EX_DBGPROP_INFO_ID  =0x0100,
    EX_DBGPROP_INFO_NTYPE  =0x0200,
@@ -68,8 +71,6 @@ enum {
    EX_DBGPROP_INFO_DEBUGEXTPROP  =0x1000
 };
 
-#endif
-
 // PythonAX Helpers.
 extern BOOL PyAXDebug_PySOURCE_TEXT_ATTR_Length(  PyObject *obAttr, ULONG *pLength );
 extern BOOL PyAXDebug_PyObject_AsSOURCE_TEXT_ATTR(  PyObject *obAttr, SOURCE_TEXT_ATTR *pstaTextAttr, ULONG numAttr );
diff --git a/setup.py b/setup.py
index 9687b21..a695c1b 100644
--- a/setup.py
+++ b/setup.py
@@ -148,7 +148,11 @@ def find_platform_sdk_dir():
     # dead ends so we only consider the job done if we find the "windows.h"
     # landmark.
     DEBUG = False # can't use log.debug - not setup yet
-    landmark = "include\\windows.h"
+    # rameshb: Newer Windows SDK have the header files in a different location
+    if get_build_version() < 11.0: # check if python was built with vs2012
+        landmark = "include\\windows.h"
+    else:
+        landmark = "include\\um\\windows.h"
     # 1. The use might have their current environment setup for the
     #    SDK, in which case the "MSSdk" env var is set.
     sdkdir = os.environ.get("MSSdk")
@@ -592,8 +596,14 @@ class WinExt_win32com_axdebug(WinExt_win32com):
     def __init__ (self, name, **kw):
         # Later SDK versions again ship with activdbg.h, but if we attempt
         # to use our own copy of that file with that SDK, we fail to link.
-        if os.path.isfile(os.path.join(sdk_dir, "include", "activdbg.h")):
-            kw.setdefault('extra_compile_args', []).append("/DHAVE_SDK_ACTIVDBG")
+        # rameshb: Newer Windows SDK have the header files in a different
+        # location
+        if get_build_version() < 11.0: # check if python was built with vs2012
+            if os.path.isfile(os.path.join(sdk_dir, "include", "activdbg.h")):
+                kw.setdefault('extra_compile_args', []).append("/DHAVE_SDK_ACTIVDBG")
+        else:
+            if os.path.isfile(os.path.join(sdk_dir, "include", "um", "activdbg.h")):
+                kw.setdefault('extra_compile_args', []).append("/DHAVE_SDK_ACTIVDBG")
         WinExt_win32com.__init__(self, name, **kw)
 
 # A hacky extension class for pywintypesXX.dll and pythoncomXX.dll
@@ -753,7 +763,12 @@ class my_build_ext(build_ext):
         assert self.compiler.initialized # if not, our env changes will be lost!
 
         is_64bit = self.plat_name == 'win-amd64'
-        extra = os.path.join(sdk_dir, 'include')
+        # rameshb: Newer Windows SDK have the header files in a different
+        # location
+        if get_build_version() < 11.0: # check if python was built with vs2012
+            extra = os.path.join(sdk_dir, 'include')
+        else:
+            extra = os.path.join(sdk_dir, 'include', 'um')
         # should not be possible for the SDK dirs to already be in our
         # include_dirs - they may be in the registry etc from MSVC, but
         # those aren't reflected here...
@@ -762,10 +777,21 @@ class my_build_ext(build_ext):
         assert os.path.isdir(extra), "%s doesn't exist!" % (extra,)
         self.compiler.add_include_dir(extra)
         # and again for lib dirs.
-        extra = os.path.join(sdk_dir, 'lib')
-        if is_64bit:
-            extra = os.path.join(extra, 'x64')
-            assert os.path.isdir(extra), extra
+        # rameshb: Newer Windows SDK have the library files in a different
+        # location
+        if get_build_version() < 11.0: # check if python was built with vs2012
+            extra = os.path.join(sdk_dir, 'lib')
+            if is_64bit:
+                extra = os.path.join(extra, 'x64')
+                assert os.path.isdir(extra), extra
+        else:
+            extra = os.path.join(sdk_dir, 'lib', 'win8', 'um')
+            if is_64bit:
+                extra = os.path.join(extra, 'x64')
+                assert os.path.isdir(extra), extra
+            else:
+                extra = os.path.join(extra, 'x86')
+                assert os.path.isdir(extra), extra
         assert extra not in self.library_dirs # see above
         assert os.path.isdir(extra), "%s doesn't exist!" % (extra,)
         self.compiler.add_library_dir(extra)
-- 
2.8.0.windows.1

